pkg/prometheus/promcfg.go

```diff
@@ -1657,7 +1659,15 @@ func (cg *ConfigGenerator) generateServiceMonitorConfig(
 	apiserverConfig *monitoringv1.APIServerConfig,
 	store *assets.StoreBuilder,
 	shards int32,
-) yaml.MapSlice {
+) (yaml.MapSlice, error) {
+
+	if m.Spec.SelectorMechanism == monitoringv1.SelectorMechanismRole && !cg.version.GTE(semver.MustParse("2.17.0")) {
+		return nil, fmt.Errorf(
+			"RoleSelector is only supported in Prometheus 2.17.0 and newer: serviceMonitor %s/%s",
```

Comment:
We should reject the resource in `*ResourceSelector.SelectServiceMonitors()`. Otherwise a single invalid ServiceMonitor would block the whole reconciliation.